When venv creator is used on Solaris (where we deliver files without
user write permission), it breaks with PermissionError.

This was reported and changes offered upstream:
https://github.com/pypa/virtualenv/issues/2419
https://github.com/pypa/virtualenv/pull/2543

From 2eabeea77bb4491be445484987065e256fdeb473 Mon Sep 17 00:00:00 2001
From: Jakub Kulik <kulikjak@gmail.com>
Date: Fri, 14 Apr 2023 15:15:24 +0200
Subject: prevent PermissionError when using venv creator on some systems

--- virtualenv-20.16.5/src/virtualenv/activation/via_template.py
+++ virtualenv-20.16.5/src/virtualenv/activation/via_template.py
@@ -37,6 +37,10 @@ class ViaTemplateActivator(Activator, me
         for template in templates:
             text = self.instantiate_template(replacements, template, creator)
             dest = to_folder / self.as_name(template)
+            # remove the file if it already exists - this prevents permission
+            # errors when the dest is not writable
+            if dest.exists():
+                dest.unlink()
             # use write_bytes to avoid platform specific line normalization (\n -> \r\n)
             dest.write_bytes(text.encode("utf-8"))
             generated.append(dest)
--- virtualenv-20.16.5/tests/unit/create/test_creator.py
+++ virtualenv-20.16.5/tests/unit/create/test_creator.py
@@ -665,3 +665,25 @@ def test_py_pyc_missing(tmp_path, mocker
 
     pyc_at = Python2.from_stdlib(Python2.mappings(CURRENT), "osc.py")[1](result.creator, Path("os.pyc"))
     assert pyc_at.exists() is pyc
+
+
+# Make sure that the venv creator works on systems where vendor-delivered files
+# (specifically venv scripts delivered with Python itself) are not writable.
+#
+# https://github.com/pypa/virtualenv/issues/2419
+@pytest.mark.skipif("venv" not in CURRENT_CREATORS, reason="test needs venv creator")
+def test_venv_creator_without_write_perms(tmp_path, mocker):
+    from virtualenv.run.session import Session
+
+    prev = Session._create
+
+    def func(self):
+        prev(self)
+        scripts_dir = self.creator.dest / "bin"
+        for script in scripts_dir.glob("*ctivate*"):
+            script.chmod(stat.S_IREAD | stat.S_IRGRP | stat.S_IROTH)
+
+    mocker.patch("virtualenv.run.session.Session._create", side_effect=func, autospec=True)
+
+    cmd = [str(tmp_path), "--seeder", "app-data", "--without-pip", "--creator", "venv"]
+    cli_run(cmd)
